src/MOI/callbacks/MOI_user_cut.jl

```diff
@@ -0,0 +1,125 @@
+            if Xpress.getintattrib(callback_data.node_model, Xpress.Lib.XPRS_CALLBACKCOUNT_OPTNODE) <= 2
+                func(callback_data)
+            end
+        end
+
+        pop_callback_state!(model)
+    end
+
+    return nothing
+end
+
+function MOI.set(model::Optimizer, ::MOI.UserCutCallback, ::Nothing)
+    info = model.callback_table.moi_user_cut
+
+    if !isnothing(info)
+        xprs_optnode_info, = info
+
+        remove_xprs_optnode_callback!(model.inner, xprs_optnode_info)
+
+        model.callback_table.moi_user_cut = nothing
+    end
+
+    return nothing
+end
+
+function MOI.set(model::Optimizer, attr::MOI.UserCutCallback, func::Function)
+    MOI.set(model, attr, nothing)
+
+    xprs_optnode_info = add_xprs_optnode_callback!(
+        model.inner,
+        (callback_data::OptNodeCallbackData) -> moi_user_cut_xprs_optnode_wrapper(
+            func,
+            model,
+            callback_data,
+        )
+    )::CallbackInfo{OptNodeCallbackData}
+
+    model.callback_table.moi_user_cut = (xprs_optnode_info,)
+
+    return nothing
+end
+
+MOI.supports(::Optimizer, ::MOI.UserCutCallback) = true
+MOI.supports(::Optimizer, ::MOI.UserCut{CD}) where {CD<:CallbackData} = true
+
+function MOI.submit(
+    model::Optimizer,
+    submittable::MOI.UserCut{CD},
+    f::F,
+    s::S,
+) where {CD<:CallbackData,T,F<:MOI.ScalarAffineFunction{T},S<:Union{MOI.LessThan{T},MOI.GreaterThan{T},MOI.EqualTo{T}}}
+    # It is assumed that every '<:CallbackData' has a 'node_model' field
+    node_model = submittable.callback_data.node_model::Xpress.XpressProblem
+
+    # TODO: Should we mark as submitted in the beginning of the function?
+    model.callback_cut_data.submitted = true
+
+    check_callback_state(model, node_model, submittable, CS_MOI_USER_CUT)
+
+    # Check if b is 0 in a'x + b <= c?
+    # f(x) = a'x + b
+    # S = {<= c}
+    if !iszero(f.constant)
+        cache_exception(
+            model,
+            MOI.ScalarFunctionConstantNotZero{T,F,S}(f.constant)
+        )
+        Xpress.interrupt(node_model, Xpress.Lib.XPRS_STOP_USER)
+
+        return nothing
+    end
+
+    indices, coefficients = _indices_and_coefficients(model, f)
+    sense, rhs = _sense_and_rhs(s)
+
+    mtype = Ref{Cint}(1) # Cut type
+    mstart = Cint[0, length(indices)]
+    mindex = Ref{Xpress.Lib.XPRScut}(C_NULL)
+    ncuts = Cint(1)
+    nodupl = Cint(2) # Duplicates are excluded from the cut pool, ignoring cut type
+    qrtype = Ref{UInt8}(sense)
+    drhs = Ref{Cdouble}(rhs)
+    indices .-= 1 # Xpress follows C's 0-index convention
+    mcols = Cint.(indices)
+    interp = Cint(-1) # Load all cuts
+    dmatval = Cdouble.(coefficients)
+
+    Xpress.Lib.XPRSstorecuts(
+        node_model,
+        ncuts,
+        nodupl,
+        mtype,
+        qrtype,
+        drhs,
+        mstart,
+        mindex,
+        mcols,
+        dmatval,
+    )
+
+    Xpress.Lib.XPRSloadcuts(
```

Comment:
User cut should not be local only?

Reply:
That lies on the second checkbox of the PR description. We are still refactoring without breaking the old API in some sense.
Implementing heuristic/lazyconstraint/usercut properly requires some redesign and other XPRS methods might be useful then.

Reply:
So you agree, but you want to fix it in a second step?

Reply:
The plan is to integrate our new callback coordination mechanism and then to make it follow MOI's semantics.
I'm adding the comments.